### PR TITLE
refactor: Upgrade Redux

### PR DESCRIPTION
**Pull request by suddjian**
### SUMMARY

At @graceguo-supercat's suggestion, I've extracted the Redux upgrade done by @pkdotson and myself in #11814. This gets us ✨ hooks ✨ for redux state!

Many tests had to be changed because Enzyme can't propagate context correctly through shallow-rendered components. Side note: can't wait to start using react-testing-library.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/actions/exploreActions.ts [hunk 1]
```diff
@@ -86,7 +86,7 @@ export function fetchFaveStar(sliceId: string) {
 
 export const SAVE_FAVE_STAR = 'SAVE_FAVE_STAR';
 export function saveFaveStar(sliceId: string, isStarred: boolean) {
-  return function (dispatch: Dispatch<ReturnType<typeof addDangerToast>>) {
```
Comment (suddjian) [inline superset-frontend/src/explore/actions/exploreActions.ts:89]: The latest version of Redux has a better-typed `dispatch` function, so the extra typing here doesn't compile anymore, for good reason.

**Review by ktmud — COMMENTED** [review]
-

**Review by ktmud — APPROVED** [review]
This is a good move!

**Review by graceguo-supercat — APPROVED** [review]
LGTM. Thanks for making this!

**Pull request merged**